rust_src/src/time.rs

```diff
@@ -354,12 +367,90 @@ pub fn float_time(time: LispObject) -> LispObject {
 
     if unsafe {
         disassemble_lisp_time(time, &mut high, &mut low, &mut usec, &mut psec) == 0
-            || decode_time_components(high, low, usec, psec, ptr::null_mut(), &mut t) == 0
+            || decode_time_components(
+                high,
+                low,
+                usec,
+                psec,
+                ExternalPtr::new(ptr::null_mut()),
+                &mut t,
+            ) == 0
     } {
         invalid_time();
     }
 
     LispObject::from_float(t)
 }
 
+fn time_add(ta: LispTime, tb: LispTime) -> LispTime {
+    let mut hi: EmacsInt = ta.hi + tb.hi;
+    let mut lo: c_int = ta.lo + tb.lo;
+    let mut us: c_int = ta.us + tb.us;
+    let mut ps: c_int = ta.ps + tb.ps;
+    us += if 1000000 <= ps { 1 } else { 0 };
+    ps -= if 1000000 <= ps { 1000000 } else { 0 };
+    lo += if 1000000 <= us { 1 } else { 0 };
+    us -= if 1000000 <= us { 1000000 } else { 0 };
+    hi += if 1 << LO_TIME_BITS <= lo { 1 } else { 0 };
+    lo -= (if 1 << LO_TIME_BITS <= lo { 1 } else { 0 }) << LO_TIME_BITS;
+    LispTime { hi, lo, us, ps }
+}
+
+fn time_subtract(ta: LispTime, tb: LispTime) -> LispTime {
+    let mut hi: EmacsInt = ta.hi - tb.hi;
+    let mut lo: c_int = ta.lo - tb.lo;
+    let mut us: c_int = ta.us - tb.us;
+    let mut ps: c_int = ta.ps - tb.ps;
+    us -= if ps < 0 { 1 } else { 0 };
+    ps += if ps < 0 { 1000000 } else { 0 };
+    lo -= if us < 0 { 1 } else { 0 };
+    us += if us < 0 { 1000000 } else { 0 };
+    hi -= if lo < 0 { 1 } else { 0 };
+    lo += (if lo < 0 { 1 } else { 0 }) << LO_TIME_BITS;
+    LispTime { hi, lo, us, ps }
+}
+
+fn time_arith<F: FnOnce(LispTime, LispTime) -> LispTime>(
+    a: LispObject,
+    b: LispObject,
+    op: F,
+) -> LispObject {
+    let mut alen: c_int = 0;
+    let mut blen: c_int = 0;
+    let ta = unsafe { lisp_time_struct(a, &mut alen) };
+    let tb = unsafe { lisp_time_struct(b, &mut blen) };
+    let t = op(ta, tb);
+    if LispObject::fixnum_overflow(t.hi) {
+        time_overflow();
+    }
+
+    let max_len = max(alen, blen);
+    let mut val = LispObject::cons(LispObject::from_C(t.ps as c_long), Qnil);
+    if max_len == 3 {
+        val = LispObject::cons(LispObject::from_C(t.us as c_long), val);
+    }
+    if max_len == 2 {
+        val = LispObject::cons(LispObject::from_C(t.lo as c_long), val);
+        val = LispObject::cons(LispObject::from_C(t.hi), val);
+    }
```

Comment:
The `switch` originally is a bit confusing, but this is not what it would convert to. In the `default` case, where length is not 2 or 3 (which I take it to mean it's actually four or more since I don't think you can actually have just 1), you still *fall through* to the `case 3` and `case 2`. The `if`s should be reading `if max_len >= 3` etc and the "first" `ps` value here should be added in case where `max_len > 3`.